File: scripts/chino/mesh.py
#!/usr/bin/env python
"""
Mesh generation
"""
import numpy as np
import cst
import meta

# metedata
dtype = meta.dtype
shape = meta.shape
delta = meta.delta
npml = meta.npml
ntop = meta.ntop

# read data
dep = np.arange( shape[2] ) * delta[2]
n = shape[:2]
x = np.fromfile( 'lon.bin', dtype ).reshape( n[::-1] ).T
y = np.fromfile( 'lat.bin', dtype ).reshape( n[::-1] ).T
z = np.fromfile( 'topo.bin', dtype ).reshape( n[::-1] ).T

# demean topography
z0 = z.mean()
z -= z0

# PML regions are extruded
for w in x, y, z:
    for i in xrange( npml, 0, -1 ):
        w[i-1,:] = w[i,:]
        w[-i,:]  = w[-i-1,:]
        w[:,i-1] = w[:,i]
        w[:,-i]  = w[:,-i-1]

# topography blending function for elevation
n = shape[2] - ntop - npml
w = 1.0 - np.r_[ np.zeros(ntop), 1.0 / (n - 1) * np.arange(n), np.ones(npml) ]

# node elevation mesh
fh = cst.util.open_excl( 'z3.bin', 'wb' )
if fh:
    for i in range( dep.size ):
        (dep[i] + z0 + w[i] * z).T.tofile( fh )
    fh.close()

# cell center locations
dep = 0.5 * (dep[:-1] + dep[1:])
x = 0.25 * (x[:-1,:-1] + x[1:,:-1] + x[:-1,1:] + x[1:,1:])
y = 0.25 * (y[:-1,:-1] + y[1:,:-1] + y[:-1,1:] + y[1:,1:])
z = 0.25 * (z[:-1,:-1] + z[1:,:-1] + z[:-1,1:] + z[1:,1:])

# topography blending function for depth
n = shape[2] - ntop - npml
w = np.r_[ np.zeros(ntop), 1.0 / n * (0.5 + np.arange(n)), np.ones(npml) ]

# write dep file
d = 'cvm/'
fh = cst.util.open_excl( d + 'dep.bin', 'wb' )
if fh:
    for i in range( dep.size ):
        (w[i] * z - dep[i]).astype( 'f' ).T.tofile( fh )
    fh.close()

# write lon file
fh = cst.util.open_excl( d + 'lon.bin', 'wb' )
if fh:
    for i in range( dep.size ):
        x.T.tofile( fh )
    fh.close()

# write lat file
fh = cst.util.open_excl( d + 'lat.bin', 'wb' )
if fh:
    for i in range( dep.size ):
        y.T.tofile( fh )
    fh.close()

File: scripts/chino/mesh.c
// Create CVM-S 3D inpute mesh from 2D mesh.
// Can be run concurrently for each file (3 processes).

#include <stdio.h>    // for printf(), fdopen, fopen(), remove()
#include <stdlib.h>   // for malloc(), exit()
#include <fcntl.h>    // for open(), mode, flags
#include <unistd.h>   // for close()
#include <string.h>   // for strrchr()
//#include <sys/stat.h>
#include "util.c"

int
main(int argc, char *argv[])
{
const float delta = DELTA;
const int nx = NX, ny = NY, nz = NZ;
const int npml = NPML, ntop = NTOP;
const int mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
const int flags = O_WRONLY | O_CREAT | O_EXCL;
const char *files[] = {"hold/lon.bin", "hold/lat.bin", "hold/dep.bin", "hold/z3.bin"};
const char *path, *path0;
const size_t nnode = nx * ny;
const size_t ncell = (nx - 1) * (ny - 1);
const size_t size = sizeof(float);
float *buff0 = (float *)malloc(nnode * sizeof(float));
float *buff = (float *)malloc(nnode * sizeof(float));
FILE *stream;
float z0, w;
int j, k, n, fh, err;
size_t i;

// lon/lat
for (k = 0; k < 2; k++) {
    path = files[k];
    fh = open(path, flags, mode);
    if (fh >= 0) {
        path0 = strrchr(path, '/') + 1;
        err = (stream = fopen(path0, "r")) == NULL;
        err = err || fread(buff, size, nnode, stream) != nnode;
        fseek(stream, 0, SEEK_END);
        err = err || ftell(stream) != size * nnode;
        if (fclose(stream) || err) {
            close(fh);
            remove(path);
            printf("Error reading %s\n", path0);
            exit(1);
        }
        extrude(buff, nx, ny, npml);
        average(buff, nx, ny);
        err = (stream = fdopen(fh, "w")) == NULL;
        for (j = 0; j < nz - 1 && !err; j++)
            err = fwrite(buff, size, ncell, stream) != ncell;
        if (fclose(stream) || err) {
            remove(path);
            printf("Error writing %s\n", path);
            exit(1);
        }
    }
}

// depth
path = files[2];
fh = open(path, flags, mode);
if (fh >= 0) {
    err = (stream = fopen("topo.bin", "r")) == NULL;
    err = err || fread(buff0, size, nnode, stream) != nnode;
    if (fclose(stream) || err) {
        close(fh);
        remove(path);
        printf("Error reading topo.bin\n");
        exit(1);
    }
    extrude(buff, nx, ny, npml);
    average(buff, nx, ny);
    err = (stream = fdopen(fh, "w")) == NULL;
    n = nz - ntop - npml - 1;
    for (j = 0; j < nz - 1 && !err; j++) {
        //XXX FIXME
        w = 0.5 + 1.0 * (j - ntop) / n;
        w = w > 0.0 ? w : 0.0;
        w = w < 1.0 ? w : 1.0;
        for (i = 0; i < nx; i++);
            buff[i] = buff[i] * w;
        err = fwrite(buff, size, ncell, stream) != ncell;
    }
    if (fclose(stream) || err) {
        remove(path);
        printf("Error writing %s\n", path);
        exit(1);
    }
}

// elevation
path = files[3];
fh = open(path, flags, mode);
if (fh >= 0) {
    err = (stream = fopen("topo.bin", "r")) == NULL;
    path0 = strrchr(path, '/') + 1;
    err = err || fread(buff0, size, nnode, stream) != nnode;
    if (fclose(stream) || err) {
        close(fh);
        remove(path);
        printf("Error reading topo.bin\n");
        exit(1);
    }
    z0 = demean(buff0, nnode);
    extrude(buff0, nx, ny, npml);
    err = (stream = fdopen(fh, "w")) == NULL;
    n = nz - ntop - npml - 1;
    for (j = 0; j < nz && !err; j++) {
        //XXX FIXME
        w = 1.0 * (j - ntop) / n;
        w = w > 0.0 ? w : 0.0;
        w = w < 1.0 ? w : 1.0;
        for (i = 0; i < nx; i++);
            buff[i] = buff[i] * w;
        err = fwrite(buff, size, nnode, stream) != nnode;
    }
    if (fclose(stream) || err) {
        remove(path);
        printf("Error writing %s\n", path);
        exit(1);
    }
}

// finished
return 0;

}

